theano/sandbox/rng_mrg.py

```diff
@@ -1077,28 +1077,77 @@ def __init__(self, seed=12345, use_cuda=None):
         self.state_updates = []
 
         super(MRG_RandomStreams, self).__init__()
+
+        # Needed to reset the streams.
+        self.default_instance_seed = seed
+
+        self.set_rstate(seed)
+
+        if use_cuda is None:
+            self.use_cuda = cuda_enabled
+        else:
+            self.use_cuda = use_cuda
+
+    def set_rstate(self, seed):
         if isinstance(seed, int):
             if seed == 0:
                 raise ValueError('seed should not be 0', seed)
             elif seed >= M2:
                 raise ValueError('seed should be less than %i' % M2, seed)
-            self.rstate = numpy.asarray([seed]*6, dtype='int32')
+            self.rstate = numpy.asarray([seed] * 6, dtype='int32')
         elif len(seed) == 6:
             if seed[0] == 0 and seed[1] == 0 and seed[2] == 0:
-                raise ValueError('The first 3 values of seed should not be all 0', seed)
+                raise ValueError(
+                    'The first 3 values of seed should not be all 0', seed)
             if seed[3] == 0 and seed[4] == 0 and seed[5] == 0:
-                raise ValueError('The last 3 values of seed should not be all 0', seed)
+                raise ValueError(
+                    'The last 3 values of seed should not be all 0', seed)
             if seed[0] >= M1 or seed[1] >= M1 or seed[2] >= M1:
-                raise ValueError('The first 3 values of seed should be less than %i' % M1, seed)
+                raise ValueError(
+                    'The first 3 values of seed should be less than %i' % M1,
+                    seed)
             if seed[3] >= M2 or seed[4] >= M2 or seed[5] >= M2:
-                raise ValueError('The last 3 values of seed should be less than %i' % M2, seed)
+                raise ValueError(
+                    'The last 3 values of seed should be less than %i' % M2,
+                    seed)
             self.rstate = numpy.asarray(seed, dtype='int32')
         else:
             raise TypeError("seed should be 1 integer or 6 integers")
-        if use_cuda is None:
-            self.use_cuda = cuda_enabled
-        else:
-            self.use_cuda = use_cuda
+
+    def seed(self, seed=None):
+        """Re-initialize each random stream
+
+        :param seed: each random stream will be assigned a unique
+        state that depends deterministically on this value.
+
+        :type seed: None or integer in range 0 to 2**30
+
+        :rtype: None
+
+        """
+        if seed is None:
+            seed = self.default_instance_seed
+        self.set_rstate(seed)
+
+        for old_r, new_r, size, nstreams in self.state_updates:
+            if nstreams is None:
+                nstreams = self.n_streams(size)
```

Comment:
if the original size is a tuple of integer vs a Theano variable, we don't have the same behavior. old_r.inputs[-1] point to the size, but as a Theano variable. If it is Constant, we don't know if the user provided a constant or if it was a tuple that we converted to Constant.

I don't want to change the old logic, so to have exactly the same behavior, I reuse the original parameter.